ratik/src/main/java/org/opencadc/ratik/ArtifactValidator.java

```diff
@@ -96,15 +120,383 @@ public ArtifactValidator(ArtifactDAO artifactDAO, URI resourceID, boolean trackS
+        // action: add siteID to Artifact.siteLocations
+        log.debug("checking explanation 4");
+        if (this.trackSiteLocations) {
+            Artifact local = this.artifactDAO.get(remote.getID());
+            if (local != null) {
+                StorageSite remoteStorageSite = getRemoteStorageSite();
+                SiteLocation remoteSiteLocation = new SiteLocation(remoteStorageSite.getID());
+                if (!local.siteLocations.contains(remoteSiteLocation)) {
+                    log.info(String.format("Artifact not in L && in R - Explanation4: Artifact in local db but "
+                                               + "siteLocations does not include remote siteID\n"
+                                               + "action: add siteID to Artifact.siteLocations\n"
+                                               + "remote - %s\nremote StorageSite - %s", remote, remoteStorageSite));
+                    //TODO do action
+                    return;
+                }
+            }
+        }
+
+    }
+
+    /**
+     * Artifact in local AND remote
+     *
+     * @param local     The local Artifact.
+     * @param remote    The remote Artifact.
+     */
+    protected void validateLocalAndRemote(Artifact local, Artifact remote) {
+        // discrepancy: artifact.uri in both && artifact.id mismatch (collision)
+        // explanation1: same ID collision due to race condition that metadata-sync has to handle
+        // evidence: no more evidence needed
+        // action: pick winner, create DeletedArtifactEvent for loser, delete loser if it is in L, insert winner if winner was in R
+        log.debug("checking artifact.id mismatch");
+        if (!local.getID().equals(remote.getID())) {
+            log.info(String.format("ID mismatch:\npick winner\nuri - %s\n local - %s %s\nremote - %s %s",
+                                   local.getURI(),
+                                   local.getID(), dateFormat.format(local.getLastModified()),
+                                   remote.getID(), dateFormat.format(remote.getLastModified())));
+            if (local.getContentLastModified().before(remote.getContentLastModified())) {
+                //TODO do action
+            }
+            return;
+        }
+
+        // discrepancy: artifact in both && valid metaChecksum mismatch
+        log.debug("checking valid metaChecksum mismatch");
+        if (!local.getMetaChecksum().equals(remote.getMetaChecksum())) {
+            log.info(String.format("MetaChecksum mismatch:\n uri - %s %s\n local - %s %s\nremote - %s %s",
+                                   local.getURI(), local.getID(), local.getMetaChecksum(),
+                                   dateFormat.format(local.getLastModified()), remote.getMetaChecksum(),
+                                   dateFormat.format(remote.getLastModified())));
+            // explanation1: pending/missed artifact update in L
```

Comment:
as above: complete the logic and log the actions taken

Reply:
Updated